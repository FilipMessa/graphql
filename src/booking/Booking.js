// @flow

import type { DepartureArrival, Leg } from '../flight/Flight';
import type { Price } from '../common/types/Price';
import type { AllowedBaggage } from './Baggage';
import type { TripData } from './types/outputs/Trip';

/**
 * This is fetched from 'bookings' endpoint.
 */
export type BookingType =
  | 'BookingOneWay'
  | 'BookingReturn'
  | 'BookingMulticity';

export type BookingsItem = {
  id: number,
  arrival: DepartureArrival,
  departure: DepartureArrival,
  legs: Array<Leg>,
  price: Price,
  created: Date,
  authToken: string,
  status: string,
  type: BookingType,
  passengerCount: number,
  trips?: TripData[],
  inbound?: TripData,
  outbound?: TripData,
  passengers: Passenger[],
};

export type TravelDocument = {|
  idNumber: string,
  expiration: number,
|};

export type Visa = {|
  info: string,
  status: string,
  timestamp: number,
  country: string,
|};

// This type could be expanded with health and passport information
export type TravelInfo = {|
  visa: Visa[],
|};

export type Pkpass = {|
  +flightNumber: string,
  +url: string,
  +passenger?: Passenger,
|};

export type Passenger = {|
  id: number,
  firstname: string,
  lastname: string,
  insuranceType: string,
  title: string,
  birthday: string,
  nationality: string,
  travelDocument: TravelDocument,
  travelInfo: TravelInfo[],
  pkpasses: $ReadOnlyArray<Pkpass> | null,
|};

export type BookedService = {|
  category: string,
  status: string,
|};

export type ContactDetails = {|
  phone: string,
  email: string,
  passenger: ?Passenger,
|};

/**
 * This is additionally fetched from 'booking/12..' endpoint if needed.
 */
export type Booking = BookingsItem & {
  allowedBaggage: AllowedBaggage,
  assets: BookingAssets,
  bookedServices: BookedService[],
  contactDetails: ContactDetails,
  allowedToChangeFlights: ?number,
  onlineCheckinIsAvailable: boolean,
  insurancePrices: InsurancePrice[],
};

export type BoardingPass = {|
  +boardingPassUrl: mixed,
  +flightNumber: string,
  +availableAt: ?string,
  +leg: ?Leg,
  +pkpasses: $ReadOnlyArray<Pkpass>,
|};

export type BookingAssets = {
  ticketUrl: ?string,
  invoiceUrl: ?string,
  boardingPasses: BoardingPass[],
  legs: Leg[],
};

export type InsurancePrice = {|
  +type: 'travel_basic' | 'travel_plus' | 'none',
  +price: Price | null,
|};
