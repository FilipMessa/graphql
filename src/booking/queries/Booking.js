// @flow

import { GraphQLID, GraphQLNonNull } from 'graphql';

import GraphQLBooking from '../types/outputs/Booking';
import type { GraphqlContextType } from '../../common/services/GraphqlContext';
import type { BookingsItem } from '../Booking';

export default {
  type: GraphQLBooking, // may be null (doesn't exist)
  description: 'Flight booking by ID.',
  args: {
    id: {
      type: new GraphQLNonNull(GraphQLID),
    },
  },
  resolve: async (
    ancestor: mixed,
    { id }: Object,
    { dataLoader }: GraphqlContextType,
  ): Promise<BookingsItem> => dataLoader.bookings.loadItem(id),
};
