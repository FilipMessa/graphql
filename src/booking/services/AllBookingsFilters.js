// @flow

import idx from 'idx';
import { DateTime } from 'luxon';

import type { BookingsItem } from '../Booking';
import type { Leg } from '../../flight/Flight';
import { CoveredBy } from '../../flight/types/enums/CoveredBy';

export function groupBookings(bookings: $ReadOnlyArray<BookingsItem>) {
  return bookings.reduce(
    (acc, curVal) => {
      if (isPastBooking(curVal) === true) {
        acc.past.push(curVal);
      } else {
        acc.future.push(curVal);
      }
      return acc;
    },
    {
      future: [],
      past: [],
    },
  );
}

export function filterOnlyBookings(
  timeFrame: 'future' | 'past',
  bookings: $ReadOnlyArray<BookingsItem>,
) {
  const bookingGroups = groupBookings(bookings);

  if (timeFrame === 'future') {
    return bookingGroups.future;
  }
  return bookingGroups.past;
}

/**
 * Past booking is booking where every arrival date of every leg is in past
 * (considering local time).
 */
export function isPastBooking(singleBooking: BookingsItem) {
  const when = singleBooking.arrival.when;
  if (!when) {
    return true;
  }

  return (
    DateTime.fromJSDate(when.utc, {
      zone: 'utc',
    })
      .diffNow('milliseconds')
      .toObject().milliseconds < 0
  );
}

export function findUpcomingLeg(
  legs: Leg[],
  guarantee: ?$Values<typeof CoveredBy>,
) {
  return [...legs]
    .filter(leg => {
      const date = idx(leg, _ => _.arrival.when.utc);
      const isFuture = date
        ? DateTime.fromJSDate(date, { zone: 'utc' }).diffNow('milliseconds')
            .milliseconds > 0
        : false;

      return guarantee ? leg.guarantee === guarantee && isFuture : isFuture;
    })
    .sort((legA, legB) => {
      const dateA = idx(legA, _ => _.arrival.when.utc);
      const dateB = idx(legB, _ => _.arrival.when.utc);

      return dateA && dateB ? dateA - dateB : 0;
    })
    .shift();
}
