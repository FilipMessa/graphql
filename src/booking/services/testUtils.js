// @flow

export const createLeg = (milliseconds: number) => ({
  arrival: {
    when: {
      utc: new Date(milliseconds),
    },
  },
});
