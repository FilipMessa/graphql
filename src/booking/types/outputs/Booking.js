// @flow

import { GraphQLEnumType, GraphQLObjectType, GraphQLString } from 'graphql';

import BookingOneWay, { type BookingOneWayData } from './BookingOneWay';
import BookingType from '../enums/BookingType';
import BookingReturn, {
  type BookingReturnData,
  splitLegs,
} from './BookingReturn';
import BookingMulticity, {
  type BookingMulticityData,
  createTrips,
} from './BookingMulticity';
import BookingDestinationImageURL from '../../resolvers/BookingDestinationImageURL';
import DeprecatedField from './Booking.deprecated';
import { commonFields } from './BookingInterface';
import type { Booking } from '../../Booking';

export default new GraphQLObjectType({
  name: 'Booking',
  description:
    'You can fetch here almost every information related to the booking itself. ' +
    'However when you want to fetch legs you have to use "oneWay", "return" and ' +
    '"multicity" fields because they highly depend on the booking type.',

  fields: {
    ...DeprecatedField,

    type: {
      type: BookingType,
      resolve: ({ type }: Booking) => type,
    },

    oneWay: {
      type: BookingOneWay,
      resolve: (booking: Booking): ?BookingOneWayData => {
        if (booking.type !== 'BookingOneWay') {
          return null;
        }

        return booking;
      },
    },

    return: {
      type: BookingReturn,
      resolve: (booking: Booking): ?BookingReturnData => {
        if (booking.type !== 'BookingReturn') {
          return null;
        }

        const { inbound, outbound } = splitLegs(booking.legs);

        return { ...booking, inbound, outbound };
      },
    },

    multicity: {
      type: BookingMulticity,
      resolve: (booking: Booking): ?BookingMulticityData => {
        if (booking.type !== 'BookingMulticity') {
          return null;
        }

        if (!Array.isArray(booking.segments)) {
          throw new Error(
            'Unexpected - booking is of type "BookingMulticity" but segments are missing.',
          );
        }

        const trips = createTrips(booking.segments, booking.legs);

        return { ...booking, trips };
      },
    },

    destinationImageUrl: {
      type: GraphQLString,
      args: {
        dimensions: {
          type: new GraphQLEnumType({
            name: 'BookingDestinationImageDimensions',
            values: {
              _1200x628: { value: '1200x628' },
              _1280x720: { value: '1280x720' },
              _220x165: { value: '220x165' },
              _275x250: { value: '275x250' },
              _300x165: { value: '300x165' },
              _375x165: { value: '375x165' },
              _600x330: { value: '600x330' },
              _600x600: { value: '600x600' },
              _610x251: { value: '610x251' },
            },
            defaultValue: '600x600',
          }),
        },
      },
      resolve: BookingDestinationImageURL,
    },

    ...commonFields,
  },
});
