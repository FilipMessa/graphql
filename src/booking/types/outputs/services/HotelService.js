// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import HotelServiceRelevantLocation from './HotelServiceRelevantLocation';

type AncestorType = {|
  +relevantLocationCodes: $ReadOnlyArray<string>,
|};

export default new GraphQLObjectType({
  name: 'HotelService',
  fields: {
    relevantLocations: {
      type: GraphQLList(HotelServiceRelevantLocation),
      resolve: ({ relevantLocationCodes }: AncestorType) =>
        relevantLocationCodes,
    },
  },
});
