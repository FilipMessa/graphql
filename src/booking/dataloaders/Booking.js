// @flow

import DataLoader from 'dataloader';
import { get } from '../../common/services/HttpRequest';
import Config from '../../../config/application';
import { sanitizeDetail } from './ApiSanitizer';

import type { Booking } from '../Booking';
import type BookingsLoader from './Bookings';

export default function createInstance(
  accessToken: ?string,
  bookingsLoader: BookingsLoader,
) {
  return new DataLoader(
    (ids: $ReadOnlyArray<number | string>) =>
      batchLoad(accessToken, bookingsLoader)(ids),
    {
      cacheKeyFn: key => parseInt(key, 10),
    },
  );
}

function batchLoad(
  accessToken: ?string,
  bookingsLoader: BookingsLoader,
): ($ReadOnlyArray<number | string>) => Promise<Array<Object>> {
  if (typeof accessToken !== 'string') {
    return () => Promise.reject(new Error('Undefined access token'));
  }
  return ids => Promise.all(ids.map(id => fetch(parseInt(id), bookingsLoader)));
}

async function fetch(bid, bookingsLoader): Promise<Booking> {
  const { authToken } = await bookingsLoader.loadItem(bid);
  const data = await get(Config.restApiEndpoint.singleBooking(bid, authToken));

  return sanitizeDetail(data);
}
