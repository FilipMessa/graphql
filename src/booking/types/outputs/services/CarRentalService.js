// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import CarRentalServiceRelevantCities, {
  type CarRentalServiceRelevantCitiesType,
} from './CarRentalServiceRelevantCities';
import { type GraphqlContextType } from '../../../../common/services/GraphqlContext';

export type CarRentalServiceType = Map<
  string,
  {|
    +pickup: Date,
    +dropoff: Date,
  |},
>;

export default new GraphQLObjectType({
  name: 'CarRentalService',
  fields: {
    relevantCities: {
      type: GraphQLList(CarRentalServiceRelevantCities),
      resolve: async (
        ancestor: CarRentalServiceType,
        args: {},
        context: GraphqlContextType,
      ): Promise<CarRentalServiceRelevantCitiesType[]> => {
        const locations = await Promise.all(
          Array.from(ancestor.keys()).map(IATA =>
            context.dataLoader.location.loadById(IATA),
          ),
        );

        return locations.map(location => {
          const ancestorValue = ancestor.get(location.locationId) || {
            pickup: new Date(),
            dropoff: new Date(),
          };

          return {
            location,
            pickup: ancestorValue.pickup,
            dropoff: ancestorValue.dropoff,
          };
        });
      },
    },
  },
});
