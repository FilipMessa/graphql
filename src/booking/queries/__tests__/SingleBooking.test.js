// @flow

import { graphql, RestApiMock } from '../../../common/services/TestingTools';
import Booking from '../Booking';
import Booking2707251 from '../../datasets/booking-2707251.json';

describe('singleBooking query', () => {
  it('should be of Booking type', () => {
    expect(Booking.type.toString()).toBe('Booking');
  });

  it('should return valid fields', async () => {
    RestApiMock.onGet(
      'https://booking-api.skypicker.com/api/v0.1/users/self/bookings/2707251\\?simple_token=[0-9a-f-]{36}',
    ).replyWithData(Booking2707251);

    const query = `
    query($id: Int!, $authToken: String!) {
      singleBooking(id: $id, authToken: $authToken) {
        id
        price{
          amount
          currency
        }
        bookingDate
        destinationImageUrl
        directAccessURL
        passengerCount
        passengers {
          nationality
        }
        bookingDate
        assets{
          ticketUrl
        }
      }
    }`;
    expect(
      await graphql(query, {
        id: 2707251,
        authToken: 'b834267c-def7-4dc8-1661-a5283d25d5e9',
      }),
    ).toMatchSnapshot();
  });
});
