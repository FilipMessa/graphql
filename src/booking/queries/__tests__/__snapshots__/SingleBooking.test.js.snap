// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`singleBooking query should return valid fields 1`] = `
Object {
  "data": Object {
    "singleBooking": Object {
      "assets": Object {
        "ticketUrl": "https://skypicker-mailing.s3.amazonaws.com/2707/1490859068_E-ticket_passenger_and_1_more_e826fc3275a65a5db8808279d8fc7f8f.pdf?v=1490859069",
      },
      "bookingDate": "2017-03-30",
      "destinationImageUrl": "https://images.kiwi.com/photos/600x600/paris_fr.grayscale.jpg",
      "directAccessURL": "https://kiwi.com/content/manage/2707251/b206db64-718f-4608-babb-0b8abe6e1b9d",
      "id": "Qm9va2luZ09uZVdheToyNzA3MjUx",
      "passengerCount": 2,
      "passengers": Array [
        Object {
          "nationality": "us",
        },
        Object {
          "nationality": "us",
        },
      ],
      "price": Object {
        "amount": 74.61,
        "currency": "USD",
      },
    },
  },
}
`;
